### ISSUE TITLE
Using slivers results in a 1-pixel gap between slivers

### ISSUE THREAD

**Issue by vanlooverenkoen** [flutter/flutter#37578]
Using slivers results in a 1-pixel gap between slivers.

You can see the 1px gap in the image below. I got this problem in every app I wrote with flutter. It is very annoying. And should really be fixed. Or maybe I am doing something wrong?

Could be related:
https://github.com/letsar/flutter_sticky_header/issues/8
https://github.com/flutter/flutter/issues/14288

This is the code used to get reproduce this issue:

```
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.black12,
      body: CustomScrollView(
        slivers: [
          SliverToBoxAdapter(
            child: Container(
              height: 150,
            ),
          ),
          SliverToBoxAdapter(
            child: Container(
              decoration: BoxDecoration(
                borderRadius: BorderRadius.only(
                  topLeft: Radius.circular(ThemeDimens.largeBorderRadius),
                  topRight: Radius.circular(ThemeDimens.largeBorderRadius),
                ),
                color: Colors.white,
              ),
              child: Padding(
                padding: const EdgeInsets.all(2),
                child: Container(
                  height: 16,
                ),
              ),
            ),
          ),
          SliverToBoxAdapter(
            child: Container(
              color: Colors.white,
              height: 5000,
            ),
          ),
        ],
      ),
    );
  }
```

<img src="https://user-images.githubusercontent.com/21172855/62423785-9704d680-b6c5-11e9-8958-bdb561ed337a.jpeg" height="550" />

**Comment by knaeckeKami** [comment]
I ran into this too.
A possible workaround for this is to use a `SliverList` with a `SliverChildListDelegate` instead of multiple `SliverToBoxAdapter`.
This prevented the 1-2 pixel gaps from occurring.
I had another issue with a pixel gap in a `SliverFillRemaining`, but I was able to work around this by using wrap the child of the SliverFillRemaining with a  `Transform.translate(offset: const Offset(0, -2) , child: ....`.

This doesn't solve the underlying issue, but may help anybody who runs into the same issue.

**Comment by vanlooverenkoen** [comment]
And does it work for all phones? Because I found that I only had this issue with android devices. but not all android devices. depends on pixel density

**Comment by knaeckeKami** [comment]
It works on all phones that I tested it, which is both iOS and Android and a range of various pixel densities. Can't guarantee it won't cause any anti aliasing artefacts for all use cases an configurations, though.

**Comment by blackraven96** [comment]
I have this problem too. Someone found a solution ?

**Comment by Liuzhiyang94** [comment]
I think this issue cause by undefine item height. if use `SliverFixedExtentList` may resolve this.

**Comment by DavidMillington-CD** [comment]
I've had this happen several times due to different things.

One time was due to a Padding widget with a top or bottom set to anything above 0.  Another was due to padding you can set on the Container widget.

The time after that it was due to the expandedHeight of the SliverAppBar.  129.0, 131.0 looks good when fully expanded but then the gap appears when scrolling.  130.0 means the gap is there from the start. 128.0 and 132.0 both have no gap when expanded or when scrolling.  Here is a Gif of me scrolling the SliverAppBar slowly to show the problem:

https://i.imgur.com/g9xbRVO.gif

Reminds me of old sub pixel rendering problems with Flash where you would get gaps and blurriness when objects were placed at non whole number locations like top: 100.5 left: 50.2

The magic number method is not a good solution as it looks good on one content page but then for another it makes the gap appear even clearer due to the difference in content height.

For now I'm cheating around this using two different methods depending on the situation.  One method is to set the background color on parents and the scaffold.  The second method is the transform trick knaeckeKami suggested.

**Comment by VladyslavBondarenko** [comment]
Reproduced with `dev v1.18.0-8.0.pre`. Not always visible with the initial code sample.

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/17024977/80591122-cd3d6f80-8a25-11ea-9400-200b6d3035f2.gif)



<details>
 <summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel dev, v1.18.0-8.0.pre, on Mac OS X 10.15.4 19E287, locale en-GB)
    • Flutter version 1.18.0-8.0.pre at /Users/nevercode/dev/flutter_dev
    • Framework revision e0c63cd35e (5 days ago), 2020-04-23 22:29:01 -0400
    • Engine revision d2ec21221e
    • Dart version 2.9.0 (build 2.9.0-3.0.dev b0d35855d8)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /Users/nevercode/Library/Android/sdk
    • Platform android-29, build-tools 29.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.4.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.4.1, Build version 11E503a
    • CocoaPods version 1.9.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 3.6)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 45.1.1
    • Dart plugin version 192.7761
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)

[✓] VS Code (version 1.44.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.9.1

[✓] Connected device (5 available)
    • Android SDK built for x86 • emulator-5554                            • android-x86    • Android 10 (API 29) (emulator)
    • Nevercode’s iPhone        • b668e524315069f3db3661ac11ff1f66afafebdb • ios            • iOS 13.4.5
    • macOS                     • macOS                                    • darwin-x64     • Mac OS X 10.15.4 19E287
    • Web Server                • web-server                               • web-javascript • Flutter Tools
    • Chrome                    • chrome                                   • web-javascript • Google Chrome 81.0.4044.129

• No issues found!
```

</details>

**Comment by niallshaw** [comment]
> And does it work for all phones? Because I found that I only had this issue with android devices. but not all android devices. depends on pixel density

@vanlooverenkoen I managed to reproduce on iPhone 11 Pro also, using v1.17.0

**Comment by balaji-ramadoss** [comment]
Any Progress in this issue?

**Comment by paulosigales** [comment]
I found a trick to hide this 1px line.
Is possible cover this gap using a BoxShadow.

```
@override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.black12,
      body: CustomScrollView(
        slivers: [
          SliverToBoxAdapter(
            child: Container(
              height: 150,
            ),
          ),
          SliverToBoxAdapter(
            child: Container(
              decoration: BoxDecoration(
                  borderRadius: BorderRadius.only(
                    topLeft: Radius.circular(20.0),
                    topRight: Radius.circular(20.0),
                  ),
                  color: Colors.white,
                  boxShadow: [
                    BoxShadow(
                        color: Colors.white,
                        blurRadius: 0.0,
                        spreadRadius: 0.0,
                        offset: Offset(0, 2)),
                  ]),
              child: Padding(
                padding: const EdgeInsets.all(2),
                child: Container(
                  height: 16,
                ),
              ),
            ),
          ),
          SliverToBoxAdapter(
            child: Container(
              color: Colors.white,
              height: 5000,
            ),
          ),
        ],
      ),
    );
  }
```

**Comment by paulosigales** [comment]
I found a trick to hide this 1px line.
Is possible cover this gap using a BoxShadow.

```
@override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.black12,
      body: CustomScrollView(
        slivers: [
          SliverToBoxAdapter(
            child: Container(
              height: 150,
            ),
          ),
          SliverToBoxAdapter(
            child: Container(
              decoration: BoxDecoration(
                borderRadius: BorderRadius.only(
                  topLeft: Radius.circular(20.0),
                  topRight: Radius.circular(20.0),
                ),
                color: Colors.white,
                boxShadow: [
                  BoxShadow(
                    color: Colors.white,
                    blurRadius: 0.0,
                    spreadRadius: 0.0,
                    offset: Offset(0, 2),
                  ),
                ],
              ),
              child: Padding(
                padding: const EdgeInsets.all(2),
                child: Container(
                  height: 16,
                ),
              ),
            ),
          ),
          SliverToBoxAdapter(
            child: Container(
              color: Colors.white,
              height: 5000,
            ),
          ),
        ],
      ),
    );
  }
```

**Comment by niallshaw** [comment]
> I found a trick to hide this 1px line.
> Is possible cover this gap using a BoxShadow.
> 
> ```
> @override
>   Widget build(BuildContext context) {
>     return Scaffold(
>       backgroundColor: Colors.black12,
>       body: CustomScrollView(
>         slivers: [
>           SliverToBoxAdapter(
>             child: Container(
>               height: 150,
>             ),
>           ),
>           SliverToBoxAdapter(
>             child: Container(
>               decoration: BoxDecoration(
>                 borderRadius: BorderRadius.only(
>                   topLeft: Radius.circular(20.0),
>                   topRight: Radius.circular(20.0),
>                 ),
>                 color: Colors.white,
>                 boxShadow: [
>                   BoxShadow(
>                     color: Colors.white,
>                     blurRadius: 0.0,
>                     spreadRadius: 0.0,
>                     offset: Offset(0, 2),
>                   ),
>                 ],
>               ),
>               child: Padding(
>                 padding: const EdgeInsets.all(2),
>                 child: Container(
>                   height: 16,
>                 ),
>               ),
>             ),
>           ),
>           SliverToBoxAdapter(
>             child: Container(
>               color: Colors.white,
>               height: 5000,
>             ),
>           ),
>         ],
>       ),
>     );
>   }
> ```



I can confirm this trick works. 
If, like in my case, you need to cover from below the SliverAppBar, ensure the following container has the same color and boxshadow color as the appBar.

```
Container(
     decoration: 
         BoxDecoration(
             color: Color(0xFF44444F),
             boxShadow: [
                  BoxShadow(
                        color: Color(0xFF44444F),
                        blurRadius: 0,
                        spreadRadius: 0,
                        offset: Offset(0, -2),
                 ),
             ],
         )
)
```

**Comment by niallshaw** [comment]
Nevermind - issue still present, fix not always working.

**Comment by Tangshenghao1** [comment]
my situation, The color of the gap is the color of the SliverAppBar, So Let me do it this way
_CustomScrollViewController.addListener((){
      if(_CustomScrollViewController.offset>180.0){
        setState(() {
          SliverAppBarBackgroundColor= Colors.orange;
        });
      }else{
        setState(() {
          SliverAppBarBackgroundColor= Colors.white;
        });
      }
    });

**Comment by TahaTesser** [comment]
<details>
<summary>code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Material App',
      theme: ThemeData.dark(),
      home: Home(),
    );
  }
}

class Home extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.black12,
      body: CustomScrollView(
        slivers: [
          SliverToBoxAdapter(
            child: Container(
              height: 150,
            ),
          ),
          SliverToBoxAdapter(
            child: Container(
              decoration: BoxDecoration(
                borderRadius: BorderRadius.only(
                  topLeft: Radius.circular(15),
                  topRight: Radius.circular(15),
                ),
                color: Colors.white,
              ),
              child: Padding(
                padding: const EdgeInsets.all(2),
                child: Container(
                  height: 16,
                ),
              ),
            ),
          ),
          SliverToBoxAdapter(
            child: Container(
              color: Colors.white,
              height: 5000,
            ),
          ),
        ],
      ),
    );
  }
}
```
</details>

<details>
<summary>flutter doctor -v</summary>

```bash
[✓] Flutter (Channel stable, 1.22.5, on macOS 11.1 20C69 darwin-x64, locale en-GB)
    • Flutter version 1.22.5 at /Users/tahatesser/Code/flutter_stable
    • Framework revision 7891006299 (6 weeks ago), 2020-12-10 11:54:40 -0800
    • Engine revision ae90085a84
    • Dart version 2.10.4

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Volumes/Extreme/SDK
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Volumes/Extreme/SDK
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Volumes/Extreme/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    • CocoaPods version 1.10.1

[!] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.52.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.18.1

[✓] Connected device (2 available)
    • RMX2001 (mobile)     • EUYTFEUSQSRGDA6D          • android-arm64 • Android 10 (API 29)
    • Taha’s iPad (mobile) • 00008020-000255113EE8402E • ios           • iOS 14.3

! Doctor found issues in 1 category.
```
```bash
[✓] Flutter (Channel master, 1.26.0-2.0.pre.419, on macOS 11.1 20C69 darwin-x64, locale en-GB)
    • Flutter version 1.26.0-2.0.pre.419 at /Users/tahatesser/Code/flutter_master
    • Framework revision 913d5933c1 (4 hours ago), 2021-01-19 22:34:03 -0800
    • Engine revision b58dbc88fc
    • Dart version 2.12.0 (build 2.12.0-240.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Volumes/Extreme/SDK
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Volumes/Extreme/SDK
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Volumes/Extreme/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.52.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.18.1

[✓] Connected device (4 available)
    • RMX2001 (mobile)     • EUYTFEUSQSRGDA6D          • android-arm64  • Android 10 (API 29)
    • Taha’s iPad (mobile) • 00008020-000255113EE8402E • ios            • iOS 14.3
    • macOS (desktop)      • macos                     • darwin-x64     • macOS 11.1 20C69 darwin-x64
    • Chrome (web)         • chrome                    • web-javascript • Google Chrome 87.0.4280.141

• No issues found!
```

</details>

**Comment by Moonspeaker** [comment]
Ran into this issue too. 
Code example: https://gist.github.com/Moonspeaker/0e8573ff6620a7e00b8f7b04937b51a1
This is how it looks on any flutter branch: https://www.youtube.com/watch?v=DDxu1NTkaMA&feature=youtu.be

**Comment by Ludonope** [comment]
@TahaTesser still found in Flutter 2.0.1

**Comment by nt4f04uNd** [comment]
I did some research and this is likely a result of anti-aliasing algorithm in skia, the layout is perfectly fine and all values passed to into painting functions are accurate.

Here's a simplified version of the reproducible

<details>
<summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(MaterialApp(home: App()));
}

class App extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.black12,
      body: Column(
        children: [
          Container(
            height: 150,
          ),
          Container(
            height: 20,
            color: Colors.white,
          ),
          Container(
            color: Colors.white,
            height: 1000,
          ),
        ],
      ),
    );
  }
}
```
</details>

**Comment by swingywc** [comment]
> I found a trick to hide this 1px line.
> Is possible cover this gap using a BoxShadow.
> 
> ```
> @override
>   Widget build(BuildContext context) {
>     return Scaffold(
>       backgroundColor: Colors.black12,
>       body: CustomScrollView(
>         slivers: [
>           SliverToBoxAdapter(
>             child: Container(
>               height: 150,
>             ),
>           ),
>           SliverToBoxAdapter(
>             child: Container(
>               decoration: BoxDecoration(
>                 borderRadius: BorderRadius.only(
>                   topLeft: Radius.circular(20.0),
>                   topRight: Radius.circular(20.0),
>                 ),
>                 color: Colors.white,
>                 boxShadow: [
>                   BoxShadow(
>                     color: Colors.white,
>                     blurRadius: 0.0,
>                     spreadRadius: 0.0,
>                     offset: Offset(0, 2),
>                   ),
>                 ],
>               ),
>               child: Padding(
>                 padding: const EdgeInsets.all(2),
>                 child: Container(
>                   height: 16,
>                 ),
>               ),
>             ),
>           ),
>           SliverToBoxAdapter(
>             child: Container(
>               color: Colors.white,
>               height: 5000,
>             ),
>           ),
>         ],
>       ),
>     );
>   }
> ```

Thanks @paulosigales, 
your solution is the best workaround to solve the gap while waiting for Google to fix this 3 years long-lasting issue (#14288).

**Comment by Piinks** [comment]
This looks like a duplicate of #14288, if so, we should close this to keep discussion and efforts focused in one place. Thanks!
